### ISSUE TITLE
performance issues with large games

### ISSUE THREAD

**Issue by sternenzaehler** [freeserf/freeserf#445]
starting from some amount of serf in game the update step starts to take more than TICK_INTERVAL milliseconds to complete. This leads to flooding the SDL Event Pipeline with timer generated update events, because they come faster than they can be processed. So the user GUI actions like clicks and keypresses are queued behind timer events in SDL event queue and are processed with a significant time lag (tens of seconds).

I have observed this on a linux machine with core I7 processor with ~16000 serfs in the game.
Below is the savegame to reproduce the issue. (just load and try any action)

The issue is could be "hacked away" by increasing the TICK_INTERVAL up to the point that it becomes responsive again.

original serf city works fine with this game.
The problem here is that this game size is by far not the largest, it just represents several hours of play with 3 AI players.

an ugly problem because it is unlikely to have a quick and elegant solution. 

a proposed not so dirty fix is stop relying on timer if the update starts taking more than TICK_INTERVAL and instead do the following pseudo code  
while (true){
check_SDL_events();
update();
}
this way the game will be responsive but will be of course slowing down as the number of serfs and buildings grows.

**Comment by wdigger** [comment]
With profiling I found that most time consuming places is object access. By now all objects links stored as objects indexes, better way - pointers is much more faster but needs too much work for implementation.
I plan this grate work for the next release.

**Comment by sternenzaehler** [comment]
You mean Collection<Flag> etc? maps of objects. So you mean accessing maps of object by objects index is most time consuming?

Why not replace map in the collection by a vector (or a good old array even). That way you wont have to change the game code, you still have object indexes and the Collection class would have the same interface as now.
The downside would be that the vector would contain empty (nullptr) elements where the object  was deleted, but it wont take more memory than a map I guess.

**Comment by wdigger** [comment]
Yap, it's about Collection.
Not sure that a sparse array will be more efficient than a map. Think, for example, about iterators.
It seems to me that we need to do it right way, but not work on temporary solution.
In any case, you may try to implement this.

**Comment by sternenzaehler** [comment]
I understand your point that sparse array is a "not so dirty hack" rather than an elegant solution.
I can try it, replace the original Collection with sparse array with some hacks and run my heavy game test.
The question is rather - would that be enough?
see, as of now it does not work with 16000 serfs. The original game works with up to 65000 serfs.
that is factor 4x. I doubt that optimizing Collection, even in your case (storing pointers to objects in the map->get_object_at_pos and in all other places) Idoubt that would give more than 2x.
It still would not be enough - as it seems.

What is your target, how many serfs you think it should handle? 
My opinion is that it would be nice to play complicated games on a 10er map with AI opponents (as soon as AI is working) and that would require at least 128 000 serfs or maybe even 256 000.
the ability to play for weeks on a large map is what makes this game unique of all settlers... all other settlers versions feature small maps and fast campaigns...

so in this case doing it "the right way" may include not only reworking the Collection but optimizing the engine as well, or multithreading it... which is bad:(

**Comment by wdigger** [comment]
When I say about pointers, I mean direct linkage between objects via pointers, without indexes at all.
Collections may be used only for iteration, list or vector it's dos no matter in this case.
This approach can cause difficalies with serialization, but it's not so critical.

**Comment by sternenzaehler** [comment]
yes I get your idea.
My question is whether that would be enough to speed the engine up to handle 65000 or more serfs.
Because anyway no matter how you store the pointers you still have to run an update() on all serfs, buildings etc during the TICK_INTERVAL.
But we will see, sparse array should be easy to implement to do a quick test.

**Comment by sternenzaehler** [comment]
many calls, yes. When I first saw the code I thought that were normal arrays, not maps. and in fact each [] operation is a map search. I was surprised.:)
but it works:) with TICK_INTERVAL 22 it works ok with 16000 serfs:) not so bad after all.
ok lets try sparse array

**Comment by sternenzaehler** [comment]
its actually already using a database lookup table - STL std::map, which has much the same principle as any database index. It does not scan the entire collection to find an index.  But this is not fast enough for the requirement to process many serfs in 20 milliseconds. There are faster methods, the fastest one is what WDIGGER recommended, but this would require changing quite a bit of code throughout all the game engine, so no quick fix. What I recommended is a quick fix that will speed things up without changing anything outside gameobject.h

**Comment by tlongstretch** [comment]
Thanks for explaining

**Comment by sternenzaehler** [comment]
not at all:) I myself just started to look at the code a couple of days ago and need lots of explaining myself)

**Comment by sternenzaehler** [comment]
Thank you for your feedback!
It is really responsive now! (But a little bit slow:) :) visibly slower than it should be)
Here it is 
21.16 : 10.98
22.64 : 0.19
24.97 : 0.18
23.53 : 0.19
22.01 : 9.70
17.83 : 0.20
23.24 : 0.19
23.32 : 0.19
24.27 : 9.64
18.23 : 0.31
24.12 : 0.17
22.33 : 0.18
25.20 : 9.63
20.76 : 0.18
21.98 : 0.19
23.70 : 0.21
23.24 : 9.05
21.67 : 0.18
22.78 : 0.17
22.95 : 0.18
23.04 : 12.13
26.07 : 0.16
24.43 : 0.18
18.71 : 0.18


Game logic is almost consistently over 20 ms, drawing time is almost zero with regular spikes up to 12 as you see. There was no scrolling or user interaction. The screen was maximized to 1920*1080, the game screen was full of action (middle of the city - to make it draw more).

But we have a new issue now - when many sounds are due at the same - the music and sounds are corrupted/garbled.  I maybe file an issue later because for now it is all testing and research.

**Comment by sternenzaehler** [comment]
What are your numbers on a Mac?

**Comment by sternenzaehler** [comment]
Hi!
Yes, I have said that above. Everything is responsive now and the game can be played.

**Comment by sternenzaehler** [comment]
WOW! factor 2x! Thank you, that is great!
11.82 : 8.58
15.08 : 0.19
13.59 : 0.19
12.71 : 0.23
16.50 : 6.38
13.55 : 0.22
12.55 : 0.17
17.32 : 0.18
12.56 : 6.63
15.00 : 0.17
13.04 : 0.19
10.61 : 0.52
12.89 : 6.28
23.79 : 0.19
14.28 : 0.18
14.55 : 0.20
16.58 : 6.29
13.54 : 0.18
12.54 : 0.15
13.87 : 0.19
14.61 : 6.28
15.18 : 0.22
Info: [main] Cleaning up...
